org.eclipse.xtext.xbase.ui/src/org/eclipse/xtext/xbase/ui/editor/XbaseEditorInputRedirector.java

```diff
@@ -0,0 +1,165 @@
+				if (fileExtensionProvider.isValid(resource.getFullPath().getFileExtension())) {
+					IJavaProject project = JavaCore.create(resource.getProject());
+					if (project.exists()) {
+						// Check if it's sitting in the output dir, look for original one
+						if (project.getOutputLocation().isPrefixOf(resource.getFullPath())) {
+							IPath relative = resource.getFullPath().removeFirstSegments(project.getOutputLocation().segmentCount());
+
+							List<IPackageFragmentRoot> packageFragmentRoots = new ArrayList<>();
+							for (IPackageFragmentRoot root : project.getPackageFragmentRoots()) {
+								if (root.getKind() == IPackageFragmentRoot.K_SOURCE) {
+									packageFragmentRoots.add(root);
+								}
+							}
+							
+							for (IPackageFragmentRoot source : packageFragmentRoots) {
+								IPath fullPath = source.getCorrespondingResource().getProjectRelativePath().append(relative);
+								IFile newFile = resource.getProject().getFile(fullPath);
+								if (newFile.exists()) {
+									return new FileEditorInput(newFile);
+								}
+							}
+						}
+
+						// Check if it's sitting in one of the output folders set on the source folders
+						List<IClasspathEntry> classpathEntries = new ArrayList<>();
+						for (IClasspathEntry classpathEntry : project.getRawClasspath()) {
+							if (classpathEntry.getEntryKind() == IClasspathEntry.CPE_SOURCE) {
+								classpathEntries.add(classpathEntry);
+							}
+						}
+						
+						for (IClasspathEntry sourceFolder : classpathEntries) {
+							if (sourceFolder.getOutputLocation() != null && sourceFolder.getOutputLocation().isPrefixOf(resource.getFullPath())) {
+								IPath relative = resource.getFullPath().removeFirstSegments(sourceFolder.getOutputLocation().segmentCount());
+								Optional<IPackageFragmentRoot> source = FluentIterable.from(project.findPackageFragmentRoots(sourceFolder))
+										.first();
+								if (source.isPresent()) {
+									IPath fullPath = source.get().getCorrespondingResource().getProjectRelativePath().append(relative);
+									IFile newFile = resource.getProject().getFile(fullPath);
+									if (newFile.exists()) {
+										return new FileEditorInput(newFile);
+									}
+								}
+							}
+						}
+
+					}
+				}
+			}
+		} catch (JavaModelException e) {
+			LOG.error("Error finding the original source for the editor input", e);
+		}
+
+		return input;
+	}
+
+	public IEditorInput findOriginalSource(IEditorInput input) {
+		IFile resource = ResourceUtil.getFile(input);
+		if (resource == null) {
+			return input;
+		}
+
+		IEditorInput original = findOriginalSourceForOuputFolderCopy(input);
+		if (original != input) {
+			return original;
+		}
+
+		IEclipseTrace trace = traceInformation.getTraceToSource(resource);
+		if (trace == null) {
+			return input;
+		}
+
+		Optional<? extends ILocationInEclipseResource> sourceInformation = Iterables.tryFind(trace.getAllAssociatedLocations(),
```

Comment:
I'd prefer a more straight forward 
```
for(PlatformResource candidate: trace.getAllAssociatedLocations()) {
  if (languageInfo.equals(candidate.getLanguage()) {
    IStorage storage = candidate.getPlatformResource();
    if (storage != null) {
      return ..
    }
  }
}
return input;
```

Reply:
The above code goes through all-the-candidates until it finds one for which `storage != null`. If you look at the original code (https://bit.ly/2wXAeXN), it gets the **first** candidate that matches the condition `languageInfo.equals(candidate.getLanguage())`, and uses it for further processing. 

@szarnekow Is it possible for more than one `candidate` to match the condition `languageInfo.equals(candidate.getLanguage())` ?